connector/routingconnector/logs.go

```diff
@@ -0,0 +1,116 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package routingconnector // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlresource"
+)
+
+type logsConnector struct {
+	component.StartFunc
+	component.ShutdownFunc
+
+	logger *zap.Logger
+	config *Config
+	router *router[consumer.Logs]
+}
+
+func newLogsConnector(
+	set connector.CreateSettings,
+	config component.Config,
+	logs consumer.Logs,
+) (*logsConnector, error) {
+	cfg := config.(*Config)
+
+	lr, ok := logs.(connector.LogsRouter)
+	if !ok {
+		return nil, errTooFewPipelines
+	}
+
+	r, err := newRouter(
+		cfg.Table,
+		cfg.DefaultPipelines,
+		lr.Consumer,
+		set.TelemetrySettings)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &logsConnector{
+		logger: set.TelemetrySettings.Logger,
+		config: cfg,
+		router: r,
+	}, nil
+}
+
+func (c *logsConnector) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (c *logsConnector) ConsumeLogs(ctx context.Context, ld plog.Logs) error {
+	// routingEntry is used to group plog.ResourceLogs that are routed to
+	// the same set of exporters.
+	// This way we're not ending up with all the logs split up which would cause
+	// higher CPU usage.
+	groups := make(map[consumer.Logs]plog.Logs)
+	var errs error
+
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		rlogs := ld.ResourceLogs().At(i)
+		rtx := ottlresource.NewTransformContext(rlogs.Resource())
+
+		noRoutesMatch := true
+		for _, route := range c.router.routes {
+			_, isMatch, err := route.statement.Execute(ctx, rtx)
+			if err != nil {
+				if c.config.ErrorMode == ottl.PropagateError {
+					return err
+				}
+				c.group(groups, c.router.defaultConsumer, rlogs)
+				continue
+			}
+			if isMatch {
+				noRoutesMatch = false
+				c.group(groups, route.consumer, rlogs)
+			}
+
+		}
+
+		if noRoutesMatch {
+			// no route conditions are matched, add resource logs to default exporters group
+			c.group(groups, c.router.defaultConsumer, rlogs)
+		}
+	}
+	for consumer, group := range groups {
```

Comment:
This confused me at first, as I didn't see groups being set, but then I realized that perhaps c.group is changing groups? If that's the case, wouldn't it be better to have c.group to return groups and set groups to it? It shouldn't cause new allocations and would make the operation more explicit.

Reply:
groups is a `map[consumer.Logs]plog.Logs`. it is mutated by `c.group`. since it is mutated by `c.group` it would be slightly confusing to me to also return and assign it. if we made that change, i would be under the impression that c.group returns a new map (without delving into the details of the method), which is not the case.

Reply:
Makes sense, thanks!